File: Blind75/91. Decode Ways/91. Decode Ways.cpp
class Solution {
public:
    map<pair<int,int>, int> memo;
    set<string> st;
    int numDecodings(string s) {
        
        for(int i=1; i<=26; i++) st.insert(to_string(i));
        if(s[0]=='0') return 0;
        return helper(s, 0, s.size()-1);
    }

    int helper(string s, int left, int right){
        if(s[left]=='0') return 0;
        if(left>=right) return 1;
        if(memo.count({left,right})) return memo[{left,right}];
        int sum = 0;
        for(int i=left; i<=right; i++){
            string curr = s.substr(left,i-left+1);
            //cout<<curr<<" "<<s.substr(i+1,right-(i+1)+1)<<endl; 
            if(st.count(curr)){
                sum+= helper(s, i+1,right);
            }
        }

        return memo[{left,right}] = sum;
    }



};

File: Blind75/91. Decode Ways/91. Decode Ways.cs
public class Solution {
    public Dictionary<string, int> number = new Dictionary<string, int>();
    public Dictionary<string, int> memo = new Dictionary<string, int>();
    public int NumDecodings(string s) {
        for(int i=1; i<=26; i++){
            number.Add(i.ToString(), 1);
        }
       return helper(s); 
    }

    public int helper(string s){
        //base case
        if(s=="") return 1;
        if(s[0]=='0') return 0;
        if(memo.ContainsKey(s)) return memo[s];
        int n = s.Length;
        int sum = 0;
        for(int i=0; i<n; i++){
            if(number.ContainsKey(s.Substring(0,i+1))){
                sum+= helper(s.Substring(i+1));
            }
        }

        return memo[s] = sum;



    }  
}
